Updating glossary and Role Management privilege documentation

SaschaPeukert opened the pull request.

SaschaPeukert committed:
Fixing headline

SaschaPeukert committed:
Updating glossary with admin commands and privileges part 1

SaschaPeukert committed:
Role management + glossary with admin commands and privileges part 2

Hunterness commented on cypher/cypher-docs/src/docs/dev/administration.asciidoc:
```diff
@@ -36,7 +36,7 @@ These include index and constraint management commands.
 
 * <<administration-security, Security>>
-** <<administration-security-introduction, User and role management>>
+** <<administration-security-introduction, Introduction>>
```
👍

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-databases-start-database, START DATABASE>>                                                                                                                       | Database administration       | Starts up a specified database.
+| <<administration-databases-stop-database, STOP DATABASE>>                                                                                                                         | Database administration       | Stops a specified database.
+| <<administration-security-roles-show, SHOW [ALL \| POPULATED] ROLES [WITH USERS]>>                                                                                                | User and role administration  | Returns information about all or populated roles, optionally including the assigned users.
+| <<administration-security-roles-create, CREATE [OR REPLACE] ROLE ... [IF NOT EXISTS] ... [AS COPY OF]>>                                                                           | User and role administration  | Creates new roles.
```
There is nothing between `IF NOT EXISTS` and `AS COPY OF` so no need for the `...` there:
`CREATE [OR REPLACE] ROLE ... [IF NOT EXISTS] [AS COPY OF]`

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-databases-show-databases, SHOW DATABASE>>                                                                                                                        | Database administration       | Returns information about a specified database.
+| <<administration-databases-show-databases, SHOW DATABASES>>                                                                                                                       | Database administration       | Returns information about all databases.
+| <<administration-databases-show-databases, SHOW DEFAULT DATABASE>>                                                                                                                | Database administration       | Returns information about the default database.
+| <<administration-databases-create-database, CREATE [OR REPLACE] DATABASE ... [IF NOT EXISTS]>>                                                                                    | Database administration       | Crates a new database.
```
`Crates` a new database

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-databases-stop-database, STOP DATABASE>>                                                                                                                         | Database administration       | Stops a specified database.
+| <<administration-security-roles-show, SHOW [ALL \| POPULATED] ROLES [WITH USERS]>>                                                                                                | User and role administration  | Returns information about all or populated roles, optionally including the assigned users.
+| <<administration-security-roles-create, CREATE [OR REPLACE] ROLE ... [IF NOT EXISTS] ... [AS COPY OF]>>                                                                           | User and role administration  | Creates new roles.
+| <<administration-security-roles-drop, DROP ROLE [IF EXISTS]>>                                                                                                                     | User and role administration  | Deletes a specified role.
```
Should you have the `...` here as well, like you have for the drop database/user?

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-roles-create, CREATE [OR REPLACE] ROLE ... [IF NOT EXISTS] ... [AS COPY OF]>>                                                                           | User and role administration  | Creates new roles.
+| <<administration-security-roles-drop, DROP ROLE [IF EXISTS]>>                                                                                                                     | User and role administration  | Deletes a specified role.
+| <<administration-security-users-show, SHOW USERS>>                                                                                                                                | User and role administration  | Returns information about all users.
+| <<administration-security-users-create, CREATE [OR REPLACE] USER ... [IF NOT EXISTS] SET PASSWORD ... [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>   | User and role administration  | Creates a new user and sets the password for the new account. Optionally the account status can also be set and if the user should change the password upon first login.
```
One more part of this command is optional, the second SET PASSWORD, and one is not, ACTIVE|SUSPENDED:
`CREATE [OR REPLACE] USER ... [IF NOT EXISTS] SET PASSWORD ... [[SET PASSWORD] CHANGE [NOT] REQUIRED] [SET STATUS {ACTIVE \| SUSPENDED}]`

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-users-show, SHOW USERS>>                                                                                                                                | User and role administration  | Returns information about all users.
+| <<administration-security-users-create, CREATE [OR REPLACE] USER ... [IF NOT EXISTS] SET PASSWORD ... [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>   | User and role administration  | Creates a new user and sets the password for the new account. Optionally the account status can also be set and if the user should change the password upon first login.
+| <<administration-security-users-drop, DROP USER ... [IF EXISTS]>>                                                                                                                 | User and role administration  | Deletes a specified user.
+| <<administration-security-users-alter, ALTER USER ... [SET PASSWORD ...] [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>                                | User and role administration  | Changes a user account. Changes can include setting a new password, setting the account status and enabling that the user should change the password upon next login.
```
Same for ACTIVE/SUSPENDED here as for CREATE USER.

When wanting to change both the password and if it needs to be changed again (since that does not happen automatically) you can skip the second `SET PASSWORD`. I don't know how we best captures it but something like `ALTER USER ... [SET PASSWORD {password [CHANGE [NOT] REQUIRED] | CHANGE [NOT] REQUIRED}] [SET STATUS {ACTIVE | SUSPENDED}]` maybe? 

I think we in most places just have `[SET PASSWORD ... [CHANGE [NOT] REQUIRED]] [SET PASSWORD CHANGE [NOT] REQUIRED]` but that would be slightly wrong since you can't have `SET PASSWORD password CHANGE REQUIRED SET PASSWORD CHANGE NOT REQUIRED`

Hunterness replied:
The documentation section about this have:

```
ALTER USER name SET {
      PASSWORD password
            [[SET PASSWORD] CHANGE [NOT] REQUIRED]
            [SET STATUS {ACTIVE | SUSPENDED} ] |
      PASSWORD CHANGE [NOT] REQUIRED
            [SET STATUS {ACTIVE | SUSPENDED}] |
      STATUS {ACTIVE | SUSPENDED}
}
```

SaschaPeukert replied:
I am going with your first option

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-users-alter-password, ALTER CURRENT USER SET PASSWORD FROM ... TO>>                                                                                     | User and role administration  | Change the password of the user that is currently logged in.
+| <<administration-security-subgraph-show, SHOW [ROLE ... \| USER ... \| ALL] PRIVILEGES>>                                                                                          | User and role administration  | Returns information about role, user or all privileges.
+| <<administration-security-roles-grant, GRANT ROLE[S] ... TO>>                                                                                                                     | Privilege administration      | Assigns one or multiple roles to one or multiple users.
+| <<administration-security-administration-database-privileges, GRANT ... ON DATABASE ... TO>>                                                                                      | Privilege administration      | Assigns a database or schema privilege to one or multiple roles.
```
database, schema or token privilege

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-administration-database-privileges, GRANT ... ON DATABASE ... TO>>                                                                                      | Privilege administration      | Assigns a database or schema privilege to one or multiple roles.
+| <<administration-security-subgraph-introduction, GRANT ... ON GRAPH ... [NODES \| RELATIONSHIPS \| ELEMENTS] ... TO>>                                                             | Privilege administration      | Assigns a graph privilege for one or multiple specified elements to one or multiple roles.
+| <<administration-security-administration-dbms-privileges-role-management, GRANT ... ON DBMS TO>>                                                                                  | Privilege administration      | Assigns a DBMS privilege to one or multiple roles.
+| <<administration-security-administration-database-privileges, DENY ... ON DATABASE ... TO>>                                                                                       | Privilege administration      | Denies a database or schema privilege to one or multiple roles.
```
database, schema or token privilege

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-subgraph-introduction, DENY ... ON GRAPH ... [NODES \| RELATIONSHIPS \| ELEMENTS] ... TO>>                                                              | Privilege administration      | Denies a graph privilege for one or multiple specified elements to one or multiple roles.
+| <<administration-security-administration-dbms-privileges-role-management, DENY ... ON DBMS TO>>                                                                                   | Privilege administration      | Denies a DBMS privilege to one or multiple roles.
+| <<administration-security-roles-revoke, REVOKE ROLE[S] ... FROM>>                                                                                                                 | Privilege administration      | Removes one or multiple roles from one or multiple users.
+| <<administration-security-administration-database-privileges, REVOKE [GRANT \| DENY] ... ON DATABASE ... FROM>>                                                                   | Privilege administration      | Removes a database or schema privilege from one or multiple roles.
```
database, schema or token privilege

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-users-show, SHOW USERS>>                                                                                                                                | User and role administration  | Returns information about all users.
+| <<administration-security-users-create, CREATE [OR REPLACE] USER ... [IF NOT EXISTS] SET PASSWORD ... [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>   | User and role administration  | Creates a new user and sets the password for the new account. Optionally the account status can also be set and if the user should change the password upon first login.
+| <<administration-security-users-drop, DROP USER ... [IF EXISTS]>>                                                                                                                 | User and role administration  | Deletes a specified user.
+| <<administration-security-users-alter, ALTER USER ... [SET PASSWORD ...] [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>                                | User and role administration  | Changes a user account. Changes can include setting a new password, setting the account status and enabling that the user should change the password upon next login.
```
I also found a small thing missing in the section about modifying a user (https://neo4j.com/docs/cypher-manual/4.0-preview/administration/security/users-and-roles/#administration-security-users-alter). We should mention something there about all other values being unaffected if not specified (for example if you have `password change required` and change only the password `ALTER USER name SET PASSWORD 'secret'` then you still have `password change required`)

Hunterness replied:
@craigtaverner said he might fix this when building and reviewing the pr

SaschaPeukert replied:
I leave this to Craig then

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-administration-database-privileges, DENY ... ON DATABASE ... TO>>                                                                                       | Privilege administration      | Denies a database or schema privilege to one or multiple roles.
+| <<administration-security-subgraph-introduction, DENY ... ON GRAPH ... [NODES \| RELATIONSHIPS \| ELEMENTS] ... TO>>                                                              | Privilege administration      | Denies a graph privilege for one or multiple specified elements to one or multiple roles.
+| <<administration-security-administration-dbms-privileges-role-management, DENY ... ON DBMS TO>>                                                                                   | Privilege administration      | Denies a DBMS privilege to one or multiple roles.
+| <<administration-security-roles-revoke, REVOKE ROLE[S] ... FROM>>                                                                                                                 | Privilege administration      | Removes one or multiple roles from one or multiple users.
```
I think assigning and revoking roles go under role management instead of privilege management

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-users-alter, ALTER USER ... [SET PASSWORD ...] [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>                                | User and role administration  | Changes a user account. Changes can include setting a new password, setting the account status and enabling that the user should change the password upon next login.
+| <<administration-security-users-alter-password, ALTER CURRENT USER SET PASSWORD FROM ... TO>>                                                                                     | User and role administration  | Change the password of the user that is currently logged in.
+| <<administration-security-subgraph-show, SHOW [ROLE ... \| USER ... \| ALL] PRIVILEGES>>                                                                                          | User and role administration  | Returns information about role, user or all privileges.
+| <<administration-security-roles-grant, GRANT ROLE[S] ... TO>>                                                                                                                     | Privilege administration      | Assigns one or multiple roles to one or multiple users.
```
I think assigning and revoking roles go under role management instead of privilege management

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-users-drop, DROP USER ... [IF EXISTS]>>                                                                                                                 | User and role administration  | Deletes a specified user.
+| <<administration-security-users-alter, ALTER USER ... [SET PASSWORD ...] [SET PASSWORD CHANGE [NOT] REQUIRED] [SET STATUS [ACTIVE \| SUSPENDED]]>>                                | User and role administration  | Changes a user account. Changes can include setting a new password, setting the account status and enabling that the user should change the password upon next login.
+| <<administration-security-users-alter-password, ALTER CURRENT USER SET PASSWORD FROM ... TO>>                                                                                     | User and role administration  | Change the password of the user that is currently logged in.
+| <<administration-security-subgraph-show, SHOW [ROLE ... \| USER ... \| ALL] PRIVILEGES>>                                                                                          | User and role administration  | Returns information about role, user or all privileges.
```
I think the `SHOW PRIVILEGES` commands go under privilege management instead of role/user management

Hunterness replied:
The `ALL` is optional, you can write just `SHOW PRIVILEGES` which defaults to `ALL`

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-administration-database-constraints, CREATE CONSTRAINT>>            | Schema                | Determines whether a user is allowed to create constraints on a specific database.
+| <<administration-security-administration-database-constraints, DROP CONSTRAINT>>              | Schema                | Determines whether a user is allowed to drop constraints on a specific database.
+| <<administration-security-administration-database-constraints, CONSTRAINT MANAGEMENT>>        | Schema                | Determines whether a user is allowed to create and drop constraints on a specific database.
+| <<administration-security-administration-database-tokens, CREATE NEW NODE LABEL>>             | Schema                | Determines whether a user is allowed to create new node labels on a specific database.
+| <<administration-security-administration-database-tokens, CREATE NEW RELATIONSHIP TYPE>>      | Schema                | Determines whether a user is allowed to create new relationship types on a specific database.
+| <<administration-security-administration-database-tokens, CREATE NEW PROPERTY NAME>>          | Schema                | Determines whether a user is allowed to create new property names on a specific database.
+| <<administration-security-administration-database-tokens, NAME MANAGEMENT>>                   | Schema                | Determines whether a user is allowed to create new labels, types and property names on a specific database.
```
These are Token privileges not Schema

SaschaPeukert replied:
"Token" is a word we use internally but it is not used in any documentation. So I am not in favor of introducing it here.

Hunterness replied:
So user NAME? The command is NAME MANAGEMENT (which I don't think falls under the SCHEMA MANAGEMENT in Ivans document (I think he might call it token there but not sure)

SaschaPeukert replied:
Still standing by my schema argument. The name of things are part of the schema in a conceptual way. Its literally what you get from calling db.schema():  the name of things (and how they are related). 
Ivans document is sadly not part of the documentation, so lets not use internal words and concepts that are not explained in the docs.

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-administration-database-tokens, CREATE NEW RELATIONSHIP TYPE>>      | Schema                | Determines whether a user is allowed to create new relationship types on a specific database.
+| <<administration-security-administration-database-tokens, CREATE NEW PROPERTY NAME>>          | Schema                | Determines whether a user is allowed to create new property names on a specific database.
+| <<administration-security-administration-database-tokens, NAME MANAGEMENT>>                   | Schema                | Determines whether a user is allowed to create new labels, types and property names on a specific database.
+| <<administration-security-administration-database-all, ALL DATABASE PRIVILEGES>>              | Database and schema   | Implies all privileges from the categories Database and Schema for a specific database.
```
Database, Schema and Token

Hunterness replied:
Changed in text but not in `Category` column

SaschaPeukert replied:
See other comment. Lets not call it Token. Let that be part of Schema

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+|===
+
+[[glossary-privileges]]
+=== Privileges
```
Should this list also mark which parts of the commands/privileges that are optional?

SaschaPeukert replied:
I think it is okay like this. Reads easier. I just did it for the other one because now it looks more similar to the Clauses table :-)

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-administration-dbms-privileges-role-management, ASSIGN ROLE>>       | DBMS                  | Determines whether the user can grant roles.
+| <<administration-security-administration-dbms-privileges-role-management, REMOVE ROLE>>       | DBMS                  | Determines whether the user can revoke roles.
+| <<administration-security-administration-dbms-privileges-role-management, SHOW ROLE>>         | DBMS                  | Determines whether the user can get information about existing and assigned roles.
+| <<administration-security-administration-dbms-privileges-role-management, ROLE MANAGEMENT>>   | DBMS                  | Determines whether the user can create, drop, grant, revoke and show roles.
```
Should this list `assign` and `remove` roles instead of `grant` and `revoke` since that is what the other privileges are called?

SaschaPeukert replied:
The way I structured this table is that I separated the name of the privilege from the "action" that can be done with it in the description. So as a user you can see that having "ASSIGN ROLE" means that you can "grant" (like the command!). That's why I would like to keep it this way

davidoliverSP2 committed:
editorial update

davidoliverSP2 committed:
update missed token privilege

davidoliverSP2 reviewed (approved):
Thanks @SaschaPeukert 

I've pushed a few editorial changes, and fixed a few of the updates requested by @Hunterness 

I'll request a review from @mariascharin as she's currently working on docs in this section. Maria is happy to merge this when all have reviewed/approved in order to ensure the minimal overlap/conflict with our ongoing 4.0 docs PRs.

SaschaPeukert committed:
Including review comments

Also: introduction of a Token concept in the glossary is not really optimal. Let's not.

Hunterness commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -287,23 +287,23 @@ The following commands are only executable against the `system` database:
-| <<administration-security-administration-database-privileges, GRANT ... ON DATABASE ... TO>>                                                                                      | Privilege administration      | Assigns a database, schema, or token privilege to one or multiple roles.
+| <<administration-security-roles-grant, GRANT ROLE[S] ... TO>>                                                                                                                     | User and role administration  | Assigns one or multiple roles to one or multiple users.
+| <<administration-security-roles-revoke, REVOKE ROLE[S] ... FROM>>                                                                                                                 | User and role administration  | Removes one or multiple roles from one or multiple users.
+| <<administration-security-subgraph-show, SHOW {ROLE ... \| USER ... \| ALL } PRIVILEGES>>                                                                                         | Privilege administration      | Returns information about role, user or all privileges.
```
Now you made it not optional... it was optional before and I missed it :P (now its a group where you should choose one)

Hunterness commented:
Then I only have the comment @craigtaverner said he might fix left :)

craigtaverner commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+| <<administration-security-users-show, SHOW USERS>>                                                                                                                                | User and role administration  | Returns information about all users.
+| <<administration-security-users-create, CREATE [OR REPLACE] USER ... [IF NOT EXISTS] SET PASSWORD ... [[SET PASSWORD] CHANGE [NOT] REQUIRED] [SET STATUS {ACTIVE \| SUSPENDED}]>> | User and role administration  | Creates a new user and sets the password for the new account. Optionally the account status can also be set and if the user should change the password upon first login.
+| <<administration-security-users-drop, DROP USER ... [IF EXISTS]>>                                                                                                                 | User and role administration  | Deletes a specified user.
+| <<administration-security-users-alter, ALTER USER ... [SET PASSWORD {password [CHANGE [NOT] REQUIRED] | CHANGE [NOT] REQUIRED}] [SET STATUS {ACTIVE | SUSPENDED}]>>               | User and role administration  | Changes a user account. Changes can include setting a new password, setting the account status and enabling that the user should change the password upon next login.
```
Link not closed!

craigtaverner replied:
Seems to be due to missing escaping of the pipe symbol. Causes all the following table rows to be misaligned.

craigtaverner replied:
Should we really use the full syntax specification here? Would it not be better and cleaner to only list enough of the command to make its purpose clear, and rely on the link to the command specification for covering the complete syntax?

craigtaverner commented on cypher/cypher-docs/src/docs/dev/glossary.asciidoc:
```diff
@@ -270,3 +272,70 @@ This section comprises a glossary of all the keywords -- grouped by category and
+|===
+
+[[glossary-privileges]]
+=== Privileges
```
Shouldn't this title be `Privilege Actions`, since the privilege is the entire permission, while the things described here are only the core `PrivilegeAction` types? A privilege itself is composed of a combination of action, resource, scope and qualifier, and applies to a database (or graph or dbms).

craigtaverner commented on cypher/cypher-docs/src/test/scala/org/neo4j/cypher/docgen/SecurityAdministrationTest.scala:
```diff
@@ -212,8 +221,86 @@ class SecurityAdministrationTest extends DocumentingTest with QueryStatisticsTes
+      }
+
       section("Using a custom role to manage DBMS privileges", "administration-security-administration-dbms-custom", "enterprise-edition") {
+        p("include::dbms/admin-role-dbms.asciidoc[]")
```
This section introduces the concept that the admin privilege cannot be built from scratch. This remains true, and is therefor better to place before the other section.

craigtaverner committed:
Updates based on code review.

craigtaverner committed:
Note on optional parts of ALTER USER

craigtaverner commented:
I have made changes based on my review, and now marked this PR as approved. There are, however, two further changes I think are worth considering, but did not want to block the PR for them:

* Make the glossary alphabetical, as all previous glossary sections were
* Simplify the glossary first column to only enough of the query to make it clear what it is, instead of the current approach of including the entire syntax. When this results in keywords being removed (eg. removal of `IF EXISTS`, then add a new row to the table with that keyword phrase and mention which commands it applies to in the description).

These changes could be done in a followup PR if deemed valuable by team-docsumentation.
